stdlib/config/config.lua

```diff
@@ -0,0 +1,152 @@
+        if c ~= nil then error("path '" .. path .. "' contains the reserved character '" .. c .. "'") end
+
+        local pathParts = path:split('.')
+        local part = config;
+        local value = nil;
+
+        for key = 1, #pathParts, 1 do
+            if (type(part) ~= "table") then
+                value = nil;
+                break;
+            end
+
+            value = part[pathParts[key]];
+            part = part[pathParts[key]];
+        end
+
+        if (type(value) == "table") then
+            --Force break references.
+            return table.deepcopy(value);
+        elseif (value ~= nil) then
+            return value;
+        else
+            return default;
+        end
+    end
+
+    --- Set a stored config value.
+    -- @param path [required] a string, config variable to set
+    -- @param data (optional) to set path to. If nil it behaves identical to Config.delete()
+    -- @return boolean true on success; false on failure.
+    function Config.set(path, data)
+        if type(path) ~= "string" or path:is_empty() then error("path is invalid") end
+
+        local config = config_table
+
+        local c = testReservedCharacters(path)
+        if c ~= nil then error("path contains the reserved character '" .. c .. "'") end
+
+        local pathParts = path:split('.')
+        local part = config;
+        local value = nil;
+
+        for key = 1, #pathParts - 1, 1 do
+            if (type(part[pathParts[key]]) ~= "table") then
+                part[pathParts[key]] = {};
+            end
+
+            value = part[pathParts[key]];
+            part = part[pathParts[key]];
+        end
+
+        part[pathParts[#pathParts]] = data;
+
+        return 1;
+    end
+
+    --- Delete a stored config value.
+    -- @param path a string, config variable to set
+    -- @return boolean, true on success, false on failure.
+    function Config.delete(path)
+        if type(path) ~= "string" or path:is_empty() then error("path is invalid") end
+
+        local config = config_table
+
+        local c = testReservedCharacters(path)
+        if c ~= nil then error("path contains the reserved character '" .. c .. "'") end
+
+        local pathParts = path:split('.')
+        local part = config
+        local value = nil
+
+        for key = 1, #pathParts - 1, 1 do
+            if (type(part[pathParts[key]]) ~= "table") then
+                return 0
+            end
+
+            part = part[pathParts[key]]
+        end
+
+        if part[pathParts[#pathParts]] == nil then
+            return 0
+        else
+            part[pathParts[#pathParts]] = nil
+            return 1
+        end
+    end
+
+    --- Test the existence of a stored config value.
+    -- @param path a string, config variable to test
+    -- @return boolean, true on success, false otherwise
+    function Config.is_set(path)
+        if type(path) ~= "string" or path:is_empty() then error("path is invalid") end
```

Comment:
Obscure lua trivia, use `error(<string>, 2)`

The second variable defaults to 1, and is how many levels up the stack the error should be sent. 1 is not correct here, because 1 level is ourself. 2 levels is the caller.

`stdlib.core.fail_if_missing` uses `3` because `1` is itself, `2` is the function doing the argument safety check in stdlib, and `3` is the caller.